README.md
# jQuery plugin para validar cédula o RUC de Ecuador

plugin para jQuery para validar la cédula o RUC de Ecuador.

****

[![Build Status](https://travis-ci.org/macool/jquery-ecuador-id-validator.png?branch=master)](https://travis-ci.org/macool/jquery-ecuador-id-validator)
[![Code Climate](https://codeclimate.com/github/macool/jquery-ecuador-id-validator.png)](https://codeclimate.com/github/macool/jquery-ecuador-id-validator)

## Cómo usar

1. jQuery `<script src='jquery.js'></script>`
2. librería JavaScript: `<script src='dist/ruc_jquery_validator.min.js'></script>`
3. seleccionar un input y aplicar el plugin:

```javascript
$("input").validarCedulaEC();
```
*Ha sido probado con jQuery 1.9.1, pero estoy casi seguro de que correrá con muchas versiones anteriores.*

## Qué hará

1. Validará el número de cédula ingresado en el input (cuando tenga 10 o 13 caracteres) al momento de dispararse un evento (por defecto change)
2. Agregará una clase al nodo del input en caso de haber una cédula incorrecta (se puede especificar qué clase agregar)
3. Tiene dos callbacks: cuando la cédula es válida y cuando es inválida. Dentro de estas funciones, `this` es el objeto jQuery en el que se disparó la validación

## Opciones y Callbacks

```javascript
var opciones = {
  strict: true,              // va a validar siempre, aunque la cantidad de caracteres no sea 10 ni 13
  events: "change",          // evento que va a disparar la validación
  the_classes: "invalid",    // clase que se va a agregar al nodo en el que se realiza la validación
  onValid: function () {},   // callback cuando la cédula es correcta.
  onInvalid: function () {}  // callback cuando la cédula es incorrecta.
};
$("input").validarCedulaEC(opciones);
```
*Dentro del contexto de las funciones anónimas del callback, `this` es un objeto javascript que referencia al nodo en el que se acaba de disparar el evento.*


## Contribuir

He utilizado [jasmine](https://github.com/pivotal/jasmine) para el testing, y está escrito en [coffeescript](http://coffeescript.org/).

## Ejemplos

Dentro de la carpeta *examples*


CHANGELOG.md
# Changelog

## v0.0.1
*Versión inicial*


spec/ruc_jquery_validatorSpec.js
(function() {
  describe("load dependences", function() {
    it("should have already loaded jQuery 1.9.1 in $", function() {
      return expect($.fn.jquery).toEqual("1.9.1");
    });
    return it("should have loaded validarCedulaEC in $.fn", function() {
      return expect($.fn.validarCedulaEC).toBeDefined();
    });
  });

  describe("RUC jQuery Validator Plugin", function() {
    var cedulaInvalida, cedulaValida;
    cedulaValida = 1104680135;
    cedulaInvalida = 1104680134;
    beforeEach(function() {
      return window.RucValidatorEc = $.fn.validarCedulaEC.RucValidatorEc;
    });
    describe("Class methods", function() {
      it("should say province code is invalid", function() {
        expect(function() {
          return new RucValidatorEc("2304680135").isValid();
        }).toThrow("Código de provincia incorrecto.");
        return expect(function() {
          return new RucValidatorEc("-204680135").isValid();
        }).toThrow("Código de provincia incorrecto.");
      });
      it("should say third digit is invalid as 7 and 8 are not allowed", function() {
        expect(function() {
          return new RucValidatorEc("1174680135").isValid();
        }).toThrow("Tercer dígito es inválido.");
        return expect(function() {
          return new RucValidatorEc("1184680135").isValid();
        }).toThrow("Tercer dígito es inválido.");
      });
      describe("tipo de cédula", function() {
        it("should say its Persona natural", function() {
          return expect(new RucValidatorEc("1104680135").validate().tipo_de_cedula).toBe("Persona natural");
        });
        it("should say its Sociedad pública", function() {
          return expect(new RucValidatorEc("1164680130001").validate().tipo_de_cedula).toBe("Sociedad pública");
        });
        return it("should say its Sociedad privada o extranjera", function() {
          return expect(new RucValidatorEc("1194680135001").validate().tipo_de_cedula).toBe("Sociedad privada o extranjera");
        });
      });
      return describe("con cédulas", function() {
        it("should say cedula is valid", function() {
          return expect(new RucValidatorEc(cedulaValida).validate().isValid()).toBeTruthy;
        });
        it("should say cedula is invalid", function() {
          return expect(new RucValidatorEc(cedulaInvalida).validate().isValid()).toBeFalsy;
        });
        it("should validate @calu 's CI", function() {
          return expect(new RucValidatorEc("1102778014").isValid()).toBeTruthy;
        });
        return it("testing some friend's CIs", function() {
          var ci, validCIs, _i, _len, _results;
          validCIs = ["1104077209", "1102077425", "1102019351"];
          _results = [];
          for (_i = 0, _len = validCIs.length; _i < _len; _i++) {
            ci = validCIs[_i];
            _results.push(expect(new RucValidatorEc(ci).isValid()).toBeTruthy);
          }
          return _results;
        });
      });
    });
    return describe("DOM behavior", function() {
      var $input;
      $input = null;
      beforeEach(function() {
        return $input = $("<input />", {
          type: "text"
        });
      });
      it("should say it's valid when the input's CI is valid", function() {
        $input.validarCedulaEC();
        $input.val(cedulaValida).trigger("change");
        return expect($input).not.toHaveClass("invalid");
      });
      it("should say it's invalid when the input's CI is invalid", function() {
        $input.validarCedulaEC();
        $input.val(cedulaInvalida).trigger("change");
        return expect($input).toHaveClass("invalid");
      });
      it("should say it's invalid when the input's CI is too short", function() {
        $input.validarCedulaEC();
        $input.val("1234").trigger("change");
        return expect($input).toHaveClass("invalid");
      });
      it("should say it's invalid when the input's CI is 11 characters long", function() {
        $input.validarCedulaEC();
        $input.val(cedulaValida.toString() + "1").trigger("change");
        return expect($input).toHaveClass("invalid");
      });
      it("should say it's invalid when the input's CI is 14 characters long", function() {
        $input.validarCedulaEC();
        $input.val(cedulaValida.toString() + "0012").trigger("change");
        return expect($input).toHaveClass("invalid");
      });
      describe("options", function() {
        it("should validate always because strict is enabled by default", function() {
          $input.validarCedulaEC();
          $input.val("1").trigger("change");
          return expect($input).toHaveClass("invalid");
        });
        it("should not validate because strict is disabled", function() {
          $input.validarCedulaEC({
            strict: false
          });
          $input.val("110468013").trigger("change");
          return expect($input).not.toHaveClass("invalid");
        });
        it("should add a class of no-valid if specified", function() {
          $input.validarCedulaEC({
            the_classes: "no-valid"
          });
          $input.val(cedulaInvalida).trigger("change");
          return expect($input).toHaveClass("no-valid");
        });
        return it("should listen for an event of blur instead of change", function() {
          $input.validarCedulaEC({
            events: "blur"
          });
          $input.val(cedulaInvalida).trigger("change");
          expect($input).not.toHaveClass("invalid");
          $input.trigger("blur");
          return expect($input).toHaveClass("invalid");
        });
      });
      return describe("callbacks", function() {
        var callback_return;
        callback_return = null;
        beforeEach(function() {
          this.callback_fn = function() {
            return callback_return = this;
          };
          return spyOn(this, "callback_fn").andCallThrough();
        });
        it("should fire a callback when CI is valid", function() {
          $input.validarCedulaEC({
            onValid: this.callback_fn
          });
          $input.val(cedulaValida).trigger("change");
          return expect(this.callback_fn).toHaveBeenCalled;
        });
        it("should fire a callback when CI is invalid", function() {
          $input.validarCedulaEC({
            onInvalid: this.callback_fn
          });
          $input.val(cedulaInvalida).trigger("change");
          return expect(this.callback_fn).toHaveBeenCalled;
        });
        it("should bind the jQuery object to the valid callback fn", function() {
          $input.validarCedulaEC({
            onValid: this.callback_fn
          });
          $input.val(cedulaValida).trigger("change");
          return expect(callback_return[0]).toBe($input[0]);
        });
        return it("should bind the jQuery object to the invalid callback fn", function() {
          $input.validarCedulaEC({
            onInvalid: this.callback_fn
          });
          $input.val(cedulaInvalida).trigger("change");
          return expect(callback_return[0]).toBe($input[0]);
        });
      });
    });
  });

}).call(this);


Gruntfile.js
module.exports = function(grunt) {
  // Project configuration.
  grunt.initConfig({
    pkg: grunt.file.readJSON('package.json'),
    coffee: {
      compile: {
        files: {
          'src/ruc_jquery_validator.js': 'src/ruc_jquery_validator.coffee',
          'spec/ruc_jquery_validatorSpec.js': 'spec/ruc_jquery_validatorSpec.coffee'
        }
      }
    },
    coffeelint: {
      lint: ['src/*.coffee', 'spec/*.coffee'],
      options: {
        'camel_case_classes': {
          level: 'ignore'
        }
      }
    },
    jasmine: {
      test: {
        src: 'src/*.js'
      },
      dist: {
        src: 'dist/*.js'
      },
      options: {
        specs: 'spec/*Spec.js',
        helpers: ['lib/jasmine-jquery-1.5.0/jasmine-jquery.js'],
        vendor: ['lib/jquery-1.9.1/jquery-1.9.1.js']
      }
    },
    watch: {
      coffee: {
        files: ['src/**/*.coffee', 'spec/*Spec.coffee'],
        tasks: ['coffeelint:lint', 'coffee:compile']
      },
      jasmine: {
        files: ['src/**/*.js', 'spec/*Spec.js'],
        tasks: 'jasmine:test'
      }
    },
    uglify: {
      options: {
        banner:
          '/**\n' +
          ' * <%= pkg.name %>\n' +
          ' * @author <%= pkg.author %>\n' +
          ' * @license <%= pkg.license %> <%= pkg.licenseURL %>\n' +
          ' * @version <%= pkg.version %> (<%= grunt.template.today("yyyy-mm-dd") %>)\n' +
          ' * @homepage <%= pkg.homepage %>\n' +
          ' */\n'
      },
      dist: {
        files: {
          'dist/ruc_jquery_validator.min.js': 'src/ruc_jquery_validator.js'
        }
      }
    }
  });

  // Load the plugins that provide tasks.
  grunt.loadNpmTasks('grunt-contrib-coffee');
  grunt.loadNpmTasks('grunt-coffeelint');
  grunt.loadNpmTasks('grunt-contrib-jasmine');
  grunt.loadNpmTasks('grunt-contrib-watch');
  grunt.loadNpmTasks('grunt-contrib-uglify');

  // Default task(s).
  grunt.registerTask('default', ['coffeelint:lint', 'coffee:compile', 'jasmine:test']);
  // Build task
  grunt.registerTask('build', ['coffee:compile', 'uglify:dist', 'jasmine:dist']);
};


javascripts/jquery.ecuador.ci.validator.0.0.1.js
/**
 * jquery-ecuador-ci-validator
 * @author Mario Andrés Correa <a@macool.me>
 * @license beer-ware license
 * @version 0.0.1 (2014-04-12)
 */
(function(){var t=function(t,i){return function(){return t.apply(i,arguments)}};(function(i){var n,e;return n=function(){function t(t){this.numero=t,this.numero=""+this.numero,this.valid=!1,this.codigo_provincia=null,this.tipo_de_cedula=null,this.already_validated=!1}return t.prototype.validate=function(){var t,i,n,e,s,r,o,a,d,h,l,u,c,p,v,f,m,_,C,g;if(10!==(C=this.numero.length)&&13!==C)throw this.valid=!1,Error("Longitud incorrecta.");if(a=22,this.codigo_provincia=parseInt(this.numero.substr(0,2),10),1>this.codigo_provincia||this.codigo_provincia>a)throw this.valid=!1,Error("Código de provincia incorrecto.");if(l=parseInt(this.numero[2],10),7===l||8===l)throw Error("Tercer dígito es inválido.");if(9===l?this.tipo_de_cedula="Sociedad privada o extranjera":6===l?this.tipo_de_cedula="Sociedad pública":6>l&&(this.tipo_de_cedula="Persona natural"),o=[],6>l)for(n=10,u=parseInt(this.numero.substr(9,1),10),s=2,g=this.numero.substr(0,9),c=0,m=g.length;m>c;c++)i=g[c],r=parseInt(i,10)*s,r>=10&&(r-=9),o.push(r),s=2===s?1:2;if(6===l){for(u=parseInt(this.numero.substr(8,1),10),n=11,e=[3,2,7,6,5,4,3,2],i=p=0;7>=p;i=++p)o[i]=parseInt(this.numero[i],10)*e[i];o[8]=0}if(9===l)for(u=parseInt(this.numero.substr(9,1),10),n=11,e=[4,3,2,7,6,5,4,3,2],i=v=0;8>=v;i=++v)o[i]=parseInt(this.numero[i],10)*e[i];for(h=0,f=0,_=o.length;_>f;f++)i=o[f],h+=i;if(d=h%n,t=0===d?0:n-d,6===l){if("0001"!==this.numero.substr(9,4))throw Error("RUC de empresa del sector público debe terminar en 0001");this.valid=t===u}if(9===l){if("001"!==this.numero.substr(10,3))throw Error("RUC de entidad privada debe terminar en 001");this.valid=t===u}if(6>l){if(this.numero.length>10&&"001"!==this.numero.substr(10,3))throw Error("RUC de persona natural debe terminar en 001");this.valid=t===u}return this},t.prototype.isValid=function(){return this.already_validated||this.validate(),this.valid},t}(),window.RucValidatorEc=n,e=function(){function e(n,e){this.$node=n,this.options=e,this.validateContent=t(this.validateContent,this),this.options=i.extend({},i.fn.validarCedulaEC.defaults,this.options),this.$node.on(this.options.events,this.validateContent)}return e.prototype.validateContent=function(){var t,i,e,s;if(e=""+this.$node.val(),t=this.options.strict,t||10!==(s=e.length)&&13!==s||(t=!0),t)try{new n(e).isValid()?(this.$node.removeClass(this.options.the_classes),this.options.onValid.call(this.$node)):(this.$node.addClass(this.options.the_classes),this.options.onInvalid.call(this.$node))}catch(r){i=r,this.$node.addClass(this.options.the_classes),this.options.onInvalid.call(this.$node)}return null},e}(),i.fn.validarCedulaEC=function(t){return this.each(function(){return new e(i(this),t)}),this},i.fn.validarCedulaEC.defaults={strict:!0,events:"change",the_classes:"invalid",onValid:function(){return null},onInvalid:function(){return null}}})(jQuery)}).call(this);


src/ruc_jquery_validator.js
(function() {
  var __bind = function(fn, me){ return function(){ return fn.apply(me, arguments); }; };

  (function($) {
    var RucValidatorEc, jQueryRucValidatorEc;
    RucValidatorEc = (function() {
      function RucValidatorEc(numero) {
        this.numero = numero;
        this.numero = this.numero.toString();
        this.valid = false;
        this.codigo_provincia = null;
        this.tipo_de_cedula = null;
        this.already_validated = false;
      }

      RucValidatorEc.prototype.validate = function() {
        var digito_verificador, i, modulo, multiplicadores, p, producto, productos, provincias, residuo, suma, tercer_digito, verificador, _i, _j, _k, _l, _len, _len1, _ref, _ref1;
        if ((_ref = this.numero.length) !== 10 && _ref !== 13) {
          this.valid = false;
          throw new Error("Longitud incorrecta.");
        }
        provincias = 22;
        this.codigo_provincia = parseInt(this.numero.substr(0, 2), 10);
        if (this.codigo_provincia < 1 || this.codigo_provincia > provincias) {
          this.valid = false;
          throw new Error("Código de provincia incorrecto.");
        }
        tercer_digito = parseInt(this.numero[2], 10);
        if (tercer_digito === 7 || tercer_digito === 8) {
          throw new Error("Tercer dígito es inválido.");
        }
        if (tercer_digito === 9) {
          this.tipo_de_cedula = "Sociedad privada o extranjera";
        } else if (tercer_digito === 6) {
          this.tipo_de_cedula = "Sociedad pública";
        } else if (tercer_digito < 6) {
          this.tipo_de_cedula = "Persona natural";
        }
        productos = [];
        if (tercer_digito < 6) {
          modulo = 10;
          verificador = parseInt(this.numero.substr(9, 1), 10);
          p = 2;
          _ref1 = this.numero.substr(0, 9);
          for (_i = 0, _len = _ref1.length; _i < _len; _i++) {
            i = _ref1[_i];
            producto = parseInt(i, 10) * p;
            if (producto >= 10) {
              producto -= 9;
            }
            productos.push(producto);
            if (p === 2) {
              p = 1;
            } else {
              p = 2;
            }
          }
        }
        if (tercer_digito === 6) {
          verificador = parseInt(this.numero.substr(8, 1), 10);
          modulo = 11;
          multiplicadores = [3, 2, 7, 6, 5, 4, 3, 2];
          for (i = _j = 0; _j <= 7; i = ++_j) {
            productos[i] = parseInt(this.numero[i], 10) * multiplicadores[i];
          }
          productos[8] = 0;
        }
        if (tercer_digito === 9) {
          verificador = parseInt(this.numero.substr(9, 1), 10);
          modulo = 11;
          multiplicadores = [4, 3, 2, 7, 6, 5, 4, 3, 2];
          for (i = _k = 0; _k <= 8; i = ++_k) {
            productos[i] = parseInt(this.numero[i], 10) * multiplicadores[i];
          }
        }
        suma = 0;
        for (_l = 0, _len1 = productos.length; _l < _len1; _l++) {
          i = productos[_l];
          suma += i;
        }
        residuo = suma % modulo;
        digito_verificador = residuo === 0 ? 0 : modulo - residuo;
        if (tercer_digito === 6) {
          if (this.numero.substr(9, 4) !== "0001") {
            throw new Error("RUC de empresa del sector público debe terminar en 0001");
          }
          this.valid = digito_verificador === verificador;
        }
        if (tercer_digito === 9) {
          if (this.numero.substr(10, 3) !== "001") {
            throw new Error("RUC de entidad privada debe terminar en 001");
          }
          this.valid = digito_verificador === verificador;
        }
        if (tercer_digito < 6) {
          if (this.numero.length > 10 && this.numero.substr(10, 3) !== "001") {
            throw new Error("RUC de persona natural debe terminar en 001");
          }
          this.valid = digito_verificador === verificador;
        }
        return this;
      };

      RucValidatorEc.prototype.isValid = function() {
        if (!this.already_validated) {
          this.validate();
        }
        return this.valid;
      };

      return RucValidatorEc;

    })();
    jQueryRucValidatorEc = (function() {
      function jQueryRucValidatorEc($node, options) {
        this.$node = $node;
        this.options = options;
        this.validateContent = __bind(this.validateContent, this);
        this.options = $.extend({}, $.fn.validarCedulaEC.defaults, this.options);
        this.$node.on(this.options.events, this.validateContent);
      }

      jQueryRucValidatorEc.prototype.validateContent = function() {
        var check, error, numero_de_cedula, _ref;
        numero_de_cedula = this.$node.val().toString();
        check = this.options.strict;
        if (!check && ((_ref = numero_de_cedula.length) === 10 || _ref === 13)) {
          check = true;
        }
        if (check) {
          try {
            if (new RucValidatorEc(numero_de_cedula).isValid()) {
              this.$node.removeClass(this.options.the_classes);
              this.options.onValid.call(this.$node);
            } else {
              this.$node.addClass(this.options.the_classes);
              this.options.onInvalid.call(this.$node);
            }
          } catch (_error) {
            error = _error;
            this.$node.addClass(this.options.the_classes);
            this.options.onInvalid.call(this.$node);
          }
        }
        return null;
      };

      return jQueryRucValidatorEc;

    })();
    $.fn.validarCedulaEC = function(options) {
      this.each(function() {
        return new jQueryRucValidatorEc($(this), options);
      });
      return this;
    };
    $.fn.validarCedulaEC.RucValidatorEc = RucValidatorEc;
    return $.fn.validarCedulaEC.defaults = {
      strict: true,
      events: "change",
      the_classes: "invalid",
      onValid: function() {
        return null;
      },
      onInvalid: function() {
        return null;
      }
    };
  })(jQuery);

}).call(this);
